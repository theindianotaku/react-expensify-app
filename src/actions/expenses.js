import database from '../firebase/firebase';

const addExpense = (expense) => ({
  type: 'ADD_EXPENSE',
  expense
});

const startAddExpense = (expenseData = {}) => {
  return (dispatch) => {
    const {
      description = '',
      note = '',
      amount = 0,
      createdAt = 0
    } = expenseData;
    const expense = { description, note, amount, createdAt };
    
    return database.ref('expenses').push(expense)
      .then((ref) => {
        dispatch(addExpense({
          id: ref.key,
          ...expense
        }));
      });
  };
};

const removeExpense = ({ id } = {}) => ({
  type: 'REMOVE_EXPENSE',
  id
});

const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
});

export { addExpense, removeExpense, editExpense, startAddExpense };
