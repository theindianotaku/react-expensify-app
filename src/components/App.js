import React, { Component } from 'react';


class App extends Component {
  render() {
    return (
      <div>
        <p>Hello there! Expensify here.</p>
      </div>
    );
  }
}

export default App;
