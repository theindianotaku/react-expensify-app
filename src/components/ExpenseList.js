import React from 'react';

const ExpenseList = () => (
  <div>
    Expense List
  </div>
);

export default ExpenseList;
