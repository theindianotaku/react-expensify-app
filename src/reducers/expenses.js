const defaultState = [];

// state = state.expenses
const expensesReducer = (state = defaultState, action) => {
  switch (action.type) {
    case 'ADD_EXPENSE':
      return [
        ...state,
        action.expense
      ];
  }
};

export default expensesReducer;
