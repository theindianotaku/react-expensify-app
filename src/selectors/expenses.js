import moment from 'moment';

const getVisibleExpenses = (
  expenses, {
    text,
    sortBy,
    startDate,
    endDate
  }
) => {
  const filteredExpense = expenses.filter((expense) => {
    const createdAtMoment = moment(expense.createdAt);
    const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true;
    const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;
    const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());

    return startDateMatch && endDateMatch && textMatch;
  });

  const sortedExpense = filteredExpense.sort((a, b) => {
    if (sortBy === 'date') {
      return a.createdAt < b.createdAt ? 1 : -1;
    } else if (sortBy === 'amount') {
      return a.amount < b.amount ? 1 : -1;
    }
    return 0;
  });

  return sortedExpense;
};

export default getVisibleExpenses;