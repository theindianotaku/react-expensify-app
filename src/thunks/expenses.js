import { addExpense } from '../actions/expenses';
import database from '../firebase/firebase';

export const startAddExpense = (expenseData = {}) => {
  return (dispatch) => {
    const {
      description = '',
      note = '',
      amount = 0,
      createdAt = 0
    } = expenseData;
    const expense = { description, note, amount, createdAt };
    
    return database.ref('expenses').push(expense)
      .then((ref) => {
        dispatch(addExpense({
          id: ref.key,
          ...expense
        }));
      });
  };
};