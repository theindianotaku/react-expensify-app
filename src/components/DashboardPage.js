import React, { Component } from 'react';

import ExpenseList from './ExpenseList';
import ExpenseListFilter from './ExpenseListFilter';
import ExpenseSummary from './ExpenseSummary';

class DashboardPage extends Component {
  render() {
    return (
      <div>
        <ExpenseSummary />
        <ExpenseListFilter />
        <ExpenseList />
      </div>
    );
  }
}

export default DashboardPage;
