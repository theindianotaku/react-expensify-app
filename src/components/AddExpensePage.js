import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';

import { startAddExpense } from '../thunks/expenses';
import ExpenseForm from './ExpenseForm';

export class AddExpensePage extends Component {
  onSubmit = (expense) => {
    this.props.startAddExpense(expense);
    this.props.history.push('/');
  };

  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <p className="page-header__title">Add Expense</p>
            <div className="page-header__actions">
              <Link className="button simple valign-wrapper" to="/dashboard">Go back</Link>
            </div>
          </div>
        </div>
        <div className="content-container m-b-l">
          <ExpenseForm
            onSubmit={this.onSubmit}
          />
        </div>
      </div>
    );
  }
}

const mapDispatchToProps = (dispatch) => ({
  startAddExpense: (expense) => dispatch(startAddExpense(expense))
});

export default connect(undefined, mapDispatchToProps)(AddExpensePage);
