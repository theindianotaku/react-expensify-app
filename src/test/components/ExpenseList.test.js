import React from 'react';
import { shallow } from 'enzyme';
import { ExpenseList } from '../../components/ExpenseList';
import expenses from '../fixtures/expenses';

test('should render ExpenseList with expenses prop', () => {
  const wrapper = shallow(<ExpenseList expenses={expenses} />);
  expect(wrapper).toMatchSnapshot();
});

test('should render ExpenseList with empty array expenses prop', () => {
  const wrapper = shallow(<ExpenseList expenses={[]} />);
  expect(wrapper).toMatchSnapshot();
});
