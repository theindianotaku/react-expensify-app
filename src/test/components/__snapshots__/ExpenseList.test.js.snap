// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseList with empty array expenses prop 1`] = `
<div>
  <p>
    No expenses
  </p>
</div>
`;

exports[`should render ExpenseList with expenses prop 1`] = `
<div>
  <ExpenseListItem
    amount={1000}
    createdAt={0}
    description="gum bills"
    id="1"
    key="1"
    note=""
  />
  <ExpenseListItem
    amount={100}
    createdAt={-345600000}
    description="rent"
    id="2"
    key="2"
    note=""
  />
  <ExpenseListItem
    amount={4550}
    createdAt={345600000}
    description="fees"
    id="3"
    key="3"
    note=""
  />
</div>
`;
